### ISSUE TITLE
[Feature]: support experimental best-effort source-map remapping in module errors

### ISSUE THREAD

**Issue by h-a-n-a** [web-infra-dev/rspack#5632]
### What problem does this feature solve?

In the current implementation of rspack, module build error and module parse error would be mapped based on the output of loader, which leads its corresponding error span not being able to mark the user input.

### What does the proposed API of configuration look like?

This does not expose any JS APIs, however it could be enabled with `experiments.rspackFuture` options. The exact option name remains to be determined.

**Comment by h-a-n-a** [comment]
## Compiler-wise

Toggle made to the whole compiler, turning in on means this will apply to every module during each compilation.

Pro:
1. Easy to use

Con:
1. The correctness of sourcemap should be guaranteed, otherwise the result will be not correct.

**Comment by hardfist** [comment]
I think this should be a opt-in feature(other than opt-out) considering we have no way to guarantee the source-map correctness and it's the same behavior like [--enable-source-maps](--enable-source-maps).
maybe we can implement it as a userland plugin first and see whether it works for most scenarios

**Comment by h-a-n-a** [comment]
> I think this should be a opt-in feature(other than opt-out) considering we have no way to guarantee the source-map correctness and it's the same behavior like [--enable-source-maps](--enable-source-maps). maybe we can implement it as a userland plugin first and see whether it works for most scenarios

I agree with you, and I would suggestion this feature should be opt-in compiler-wise or module rule wise. But feel free to propose with a better solution if you have any ;-).

**Comment by hardfist** [comment]
```
Edit: Plugin looks good to me too, but it does not solve the problem of compiler-wise or module rule wise.
```
It seems you can control this in plugin options? since you got whole sourcemap and error location info, so you can do reverse mapping for any module?

**Comment by h-a-n-a** [comment]
> ```
> Edit: Plugin looks good to me too, but it does not solve the problem of compiler-wise or module rule wise.
> ```
> 
> It seems you can control this in plugin options? since you got whole sourcemap and error location info in stats, so you can do reverse mapping for any module?

Sure we have the capability to handle this, the design part should be deliberately considered for the best user experience.

**Comment by h-a-n-a** [comment]
bump
